### ISSUE TITLE
Advanced Indexing CPU performance regression from 1.13.1 onwards

### ISSUE THREAD

**Issue by JensWehner** [pytorch/pytorch#104028]
### 🐛 Describe the bug

In the case of advanced indexing for reading or writing from tensors we found in some cases 20-30% runtime increases with newer torch versions. This was reproducible on 2 different x86_64 and one arm machine. 
![min_arm_py3 9 16 cpu_getitem_read](https://github.com/pytorch/pytorch/assets/12139113/6988b029-d160-44e3-9671-34f33061d55e)
![min_x86_64_py3 8 16 cpu_rand_write](https://github.com/pytorch/pytorch/assets/12139113/9327ec2e-ca7b-44b1-b980-95d87bd07ebd)

The python benchmark is attached. Each test was run more than 10k times (after a warmup of 10k iterations) and the minimum from each run was selected. All tests are basically indexing operations into arrays for reading or writing. In comparison we also tested the same operations using numpy. Tensor size is (1000,) and a random bitmask with 500 true entries was generated, which was used for all calculations. All torch versions were simply installed via pip.

https://gist.github.com/JensWehner/3524a121f42a3ec912fa8f843249b75d

Code can be run via 
`pytest --randomly-seed=1 --benchmark-verbose --benchmark-only --benchmark-disable-gc --benchmark-warmup=on --benchmark-warmup-iterations=10000 -m torchperformance test_performance_tensor_access.py`

### Versions

```
NVIDIA RTX A3000 Laptop GPU with CUDA capability sm_86 is not compatible with the current PyTorch installation.
The current PyTorch install supports CUDA capabilities sm_37 sm_50 sm_60 sm_70.
If you want to use the NVIDIA RTX A3000 Laptop GPU GPU with PyTorch, please check the instructions at https://pytorch.org/get-started/locally/

  warnings.warn(incompatible_device_warn.format(device_name, capability, " ".join(arch_list), device_name))
PyTorch version: 1.10.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.9.17 (main, Jun  6 2023, 20:11:21)  [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA RTX A3000 Laptop GPU
Nvidia driver version: 517.51
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   39 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          6
On-line CPU(s) list:             0-5
Vendor ID:                       GenuineIntel
Model name:                      11th Gen Intel(R) Core(TM) i5-11500H @ 2.90GHz
CPU family:                      6
Model:                           141
Thread(s) per core:              1
Core(s) per socket:              6
Socket(s):                       1
Stepping:                        1
BogoMIPS:                        5836.79
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm avx512_vp2intersect md_clear flush_l1d arch_capabilities
Hypervisor vendor:               Microsoft
Virtualization type:             full
L1d cache:                       288 KiB (6 instances)
L1i cache:                       192 KiB (6 instances)
L2 cache:                        7.5 MiB (6 instances)
L3 cache:                        12 MiB (1 instance)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] flake8==3.9.2
[pip3] flake8-bugbear==21.4.3
[pip3] flake8-builtins==1.5.3
[pip3] flake8-comprehensions==3.5.0
[pip3] flake8-docstrings==1.6.0
[pip3] flake8-fixme==1.1.1
[pip3] flake8-isort==4.0.0
[pip3] flake8-logging-format==0.6.0
[pip3] flake8-polyfill==1.0.2
[pip3] mypy==1.3.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] torch==1.10.1
[pip3] triton==2.0.0
```

```
Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.4 (arm64)
GCC version: Could not collect
Clang version: 16.0.1
CMake version: version 3.26.3
Libc version: N/A

Python version: 3.9.16 (main, Jan 16 2023, 16:30:11)  [Clang 15.0.3 ] (64-bit runtime)
Python platform: macOS-13.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] torch==2.0.1
[conda] No relevant packages
```

```
Collecting environment information...
PyTorch version: 1.10.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Amazon Linux 2 (x86_64)
GCC version: (GCC) 7.3.1 20180712 (Red Hat 7.3.1-15)
Clang version: 11.1.0 (Amazon Linux 2 11.1.0-1.amzn2.0.2)
CMake version: version 3.26.3
Libc version: glibc-2.26

Python version: 3.8.16 (default, Mar 10 2023, 03:07:48)  [GCC 7.3.1 20180712 (Red Hat 7.3.1-15)] (64-bit runtime)
Python platform: Linux-4.14.314-238.539.amzn2.x86_64-x86_64-with-glibc2.2.5
Is CUDA available: True
CUDA runtime version: 11.2.152
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 460.32.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               79
Model name:          Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
Stepping:            1
CPU MHz:             2700.303
CPU max MHz:         3000.0000
CPU min MHz:         1200.0000
BogoMIPS:            4600.00
Hypervisor vendor:   Xen
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            46080K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt

Versions of relevant libraries:
[pip3] flake8==3.9.2
[pip3] flake8-bugbear==21.4.3
[pip3] flake8-builtins==1.5.3
[pip3] flake8-comprehensions==3.5.0
[pip3] flake8-docstrings==1.6.0
[pip3] flake8-fixme==1.1.1
[pip3] flake8-isort==4.0.0
[pip3] flake8-logging-format==0.6.0
[pip3] flake8-polyfill==1.0.2
[pip3] mypy==1.3.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] torch==1.10.1
[pip3] triton==2.0.0
[conda] Could not collect
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

**Comment by mingfeima** [comment]
@CaoE could you please follow up this issue ?

**Comment by CaoE** [comment]
I will follow up this issue.

**Comment by CaoE** [comment]
@JensWehner I can not run https://gist.github.com/JensWehner/3524a121f42a3ec912fa8f843249b75d by the command `pytest --randomly-seed=1 --benchmark-verbose --benchmark-only --benchmark-disable-gc --benchmark-warmup=on --benchmark-warmup-iterations=10000 -m torchperformance test_performance_tensor_access.py`
May I know the version of pytest you are using.

**Comment by JensWehner** [comment]
Hi @CaoE what is the error. We use the newest release of pytest. I forgot thst you also need pytest-benchmark and pytest-randomly, tjey are plugins for pytest. Sorry for the incomplete info

**Comment by CaoE** [comment]
@JensWehner I used a simple benchmark for `write_indices` with `setitem` shown as below, and I also enabled jemalloc by `export LD_PRELOAD=${CONDA_PREFIX}/lib/libjemalloc.so` and `export MALLOC_CONF="oversize_threshold:1,background_thread:true,metadata_thp:auto,dirty_decay_ms:-1,muzzy_decay_ms:-1`
```
import torch
import time

tsize = 1000
torch.manual_seed(37)

a = torch.randn(tsize, dtype=torch.float)

b = torch.arange(0, tsize - 1, dtype=torch.long)
values = torch.randn(b.shape, dtype=torch.float)

# warm up
for i in range(5000):
    a[b] = values

begin = time.time()
for i in range(10000):
    a[b] = values

end = time.time()

print("Time : ",  (end - begin) * 1000.0 / 10000)
```
I got very close performance:

```
pytorch 1.11: 0.00561 ms
pytorch 2.01: 0.00564 ms
```

Even if do not use jemalloc:
```
pytorch 1.11: 0.00545 ms
pytorch 1.11: 0.00556 ms
pytorch 1.11: 0.00551 ms

pytorch 2.01:  0.00542 ms
pytorch 2.01:  0.00533 ms
pytorch 2.01:  0.00531 ms
```

**Comment by JensWehner** [comment]
Thanks, that is interesting. I do not have access to my pc this week. (Vacation) I will have a look next week.

Cheers

**Comment by JensWehner** [comment]
@CaoE 

So I admit it is quite hard to get useful benchmarks for these operations. When I reduced the tensor size to 100 the signal becomes much clearer. 

```python
import torch
import time

tsize = 100
torch.manual_seed(37)
torch.set_num_threads(1)
repeats = 10_000_000
print("Torch version:",torch.__version__)
a = torch.randn(tsize, dtype=torch.float)
b = torch.arange(0, tsize - 1, dtype=torch.long)
values = torch.randn(b.shape, dtype=torch.float)

# warm up
for i in range(10_000):
    a[b] = values

begin = time.time()
for i in range(repeats):
    a[b] = values

end = time.time()

print("Time : ",  (end - begin) * 1000.0 / repeats)
```

```
Torch version: 2.0.1+cu117
Time :  0.0023663573265075683
```
```
Torch version: 1.10.1+cu102
Time :  0.0021014241695404053
```

**Comment by CaoE** [comment]
@JensWehner 
For `write_indices` with `setitem`, size100 may be too small (considering the cost of Python ->C++ and dispatch). 
I'm afraid other ops on such small size may also be slower compared to older versions.
For index_select , could you please provide more numbers of different problem sizes, so we can see comprehensive information on problem sizes changes.

**Comment by JensWehner** [comment]
@CaoE did sth. change in the dispatching in the last versions?


For `index_select`:
```
Torch version: 1.10.1+cu102
Size: 100 Time :  0.001133955979347229
Size: 500 Time :  0.0014111913681030275
Size: 1000 Time :  0.0018764878988265991
Size: 5000 Time :  0.004279256534576416
Size: 10000 Time :  0.007357325434684753
Size: 50000 Time :  0.03366307039260864
```

```
Torch version: 2.0.1+cu117
Size: 100 Time :  0.0015400503873825074
Size: 500 Time :  0.0017761608123779296
Size: 1000 Time :  0.002016613554954529
Size: 5000 Time :  0.0043136533260345455
Size: 10000 Time :  0.007273420357704162
Size: 50000 Time :  0.030521628952026366
```

Seems like for larger tensors the effect is gone.

**Comment by CaoE** [comment]
@JensWehner Thanks for the benchmark numbers. As mentioned before, when size is small the overhead of Python <->C++ and dispatch has a significant impact. From version 1.10 to version 2.0, there should be many changes in PyTorch core. I am not sure which specific changes have changed the overhead. This overhead will decrease as the size increases, and may has little impact on the end-to-end performance.

**Comment by CaoE** [comment]
With pleasure. If you have any questions, you can reopen it at any time.
